# Extending Trial division

## Interactive article
The goal of this jupyter notebook is to give you, the reader, some tools and additional insights in the extended sieve.
The examples used in the text are here illustrated for you by code.

It is possible to change code and run it as you wish. 
First the function is defined as well as a simple implementation that checks if something is prime.

You can change C from 41 to 61 or any other number to discover if everything still works or not.

In [10]:
C= 41
def f(x, y):
    """ Definition of the function used of the form f(x, y) = x^2 +x*y + C*y^2 """
    return x**2+x*y+C*y**2


def is_prime(N):
    """ Checks if a number is prime by simple trial division """
    x = 2  # initialize variable
    while x <= int(N ** 0.5):  # while x is smaller than the square root of N
        if N % x == 0:  # if N is divisible by x
            return False  # then N is not a prime
        x += 1  # add one more to x to test the next number
    return True  # when no x can be found that divides N, then N must be a prime number

The small subset of integers generated by f(x, y) = x^2+x*y+C*y^2. Below a sample.

In [11]:
for x in range(0,8):
    for y in range(1,2):
       print(f(x,y))

41
43
47
53
61
71
83
97


Some examples are given here below, which will be used later in the text.

In [12]:
examples = [[2, 1], [9,1], [11, 12], [11, 13]]
for example in examples:
    print(f"({example[0]}, {example[1]}) = {f(example[0], example[1])}")


(2, 1) = 47
(9, 1) = 131
(11, 12) = 6157
(11, 13) = 7193


<b>Example Observation 1:</b> <br>
Consider (11,12) = 6157, which can be factored into 47 and 131.

In [13]:
x, y= 11, 12
print(f"({x}, {y}) = {f(x, y)}")
print(f"({x}, {y}) = 47 * 131= {47*131}")
print(f"({x}, {y}) = (2,1) * (9,1)= {f(2,1) * f(9,1)}")

(11, 12) = 6157
(11, 12) = 47 * 131= 6157
(11, 12) = (2,1) * (9,1)= 6157


<b>Example Observation 2:</b> <br>
Consider (11, 13) = 7193, for which no factors can be found. 

In [14]:
x, y= 11, 13
print(f"({x}, {y}) = {f(x, y)}")
print(f"({x}, {y}) => No factorization possible (is_prime: {is_prime(f(x,y))})")
print(f"7193 => No factorization possible (is_prime: {is_prime(f(x,y))})")


(11, 13) = 7193
(11, 13) => No factorization possible (is_prime: True)
7193 => No factorization possible (is_prime: True)



## What happens if C=61 ?

In [15]:
C= 61
def f(x, y):
    """ Definition of the function used of the form f(x, y) = x^2 +x*y + C*y^2 """
    return x**2+x*y+C*y**2

The small subset of integers generated by f(x, y) = x^2+x*y+C*y^2. Below a sample.

In [16]:
for x in range(0,8):
    for y in range(1,2):
       print(f(x,y))

61
63
67
73
81
91
103
117


<b>Example Observation:</b> <br>
Consider (5, 1) = 91. 

In [17]:
x, y= 5, 1
print(f"({x}, {y}) = {f(x, y)}")
print(f"{7*13} = 7 * 13")
print(f"({x}, {y}) => No factorization possible in (x, y), despite not being prime(is_prime: {is_prime(f(x,y))})")




(5, 1) = 91
91 = 7 * 13
(5, 1) => No factorization possible in (x, y), despite not being prime(is_prime: False)


<b>Example Observation:</b> <br>
Consider (81, 7) = 10117. 

In [18]:
x, y= 81, 7
print(f"({x}, {y}) = {f(x, y)}")
print(f"({x}, {y}) = 67 * 151= {67*151}")
print(f"({x}, {y}) = (2,1) * (9,1)= {f(2,1) * f(9,1)}")


(81, 7) = 10117
(81, 7) = 67 * 151= 10117
(81, 7) = (2,1) * (9,1)= 10117
